fetch latest dotfiles

tomhoover opened the pull request:
If ~/.dotfiles already exists, pull latest changes before
executing 'script/setup' and 'script/bootstrap'.

closes #96

MikeMcQuaid commented:
My concern with this is that it won't behave appropriately if there's local, uncommitted changes to the dotfiles. What do you think would be best to do in that situation? `git pull --rebase` with autostash enabled seems like it could be a happy middle ground.

tomhoover commented:
I can see a couple of different ways to address it:

1. Use `git pull --rebase --autostash` instead of `git pull`.
2. Execute `git config pull.rebase true; git config rebase.autoStash true` on the local repository, which would allow a simple `git pull` to work.
3. As a `git pull` should fail if there are local uncommitted changes to the dotfiles, add a note to the documentation that the user manually execute a `git config pull.rebase true; git config rebase.autoStash true` on their local repository only IF they wish to enable the rebase/autostash feature. 

Number 1 is the cleanest and number 3 is the most flexible. I guess another option would be to prompt the user during the initial run of the script whether or not to write the configs in number 2--subsequent runs of the script would then carry out their chosen option automatically.

My recommendation is for me to incorporate number 1; however, can you think of any reason why either of the other options might be preferable?

tomhoover commented:
@MikeMcQuaid, any hints on correcting the travis-ci failure? The only change between my proposed commit (which failed), and your last merge (which passed) are the following lines:

    else
      (
        cd ~/.dotfiles
        git pull $Q
      )

I don't understand how this change could cause the test to fail. I'm new to travis-ci, but it appears the `spellcheck strap.sh` failed on line 63

    In strap.sh line 63:
    logn()  { STRAP_STEP="$*"; printf -- "--> $* "; }
                                     ^-- SC2059: Don't use variables in the printf format string. Use printf "..%s.." "$foo".
    The command "./script/cibuild" exited with 1.

 that had passed your previous build?? I'd be happy to correct items that are causing `spellcheck` to squawk, but I'm trying to understand why there's an inconsistent failure (it's probably just my lack of understanding) ☹️

MikeMcQuaid commented:
I'd prefer number 1, thanks. I'm sure you haven't caused the test failure, it's like just a newer, stricter version of `shellcheck` and I appreciate you fixing it up 🙇

tomhoover commented:
@MikeMcQuaid, I had to "poke" Travis to make him run again, as the previous test timed out. It appears I've finally made him happy! 😄

tomhoover commented:
I went ahead and squashed the multiple commits into two clean commits.

tomhoover commented:
Another timeout on the build--I've poked Travis again.

tomhoover commented:
Three out of the last four attempts to test have failed with travis-ci time-outs while executing the line: `sudo rm -rf /usr/local/bin/brew /usr/local/.??* /Applications/Xcode.app /Library/Developer/CommandLineTools /opt/homebrew-cask/Caskroom`.

A review of the two most recent successful tests revealed this line took 539 seconds and 547 seconds to execute. https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received states that travis-ci will fail with a time-out after 600 seconds without receiving an output from the command; therefore, even the passing tests only had 21 seconds and 13 seconds to spare before they would have timed out.

I've modified the line in question by prefixing the command with `travis_wait`, which will double the allowable time before failing with a time-out. https://www.traviscistatus.com/ shows that travis-ci has been suffering from slow OS X builds all day today, but the `travis_wait` should take care of any intermittent slowing. 

Additionally, I noticed the command line included the deprecated `/opt/homebrew-cask/Caskroom` rather than the current `/usr/local/Caskroom`--this was also corrected.

I've tested with travis-ci several times on my fork of the strap repository, and things seem to be working well. I'll push the updated commits to this PR in a moment--hopefully it continues to pass the tests.

tomhoover commented:
@MikeMcQuaid, it appears things are working now. Time for bed...Goodnight...

MikeMcQuaid commented:
Great work, thanks @tomhoover!
